pkg/fqdn/dnspoller.go

```diff
@@ -0,0 +1,381 @@
+			// was set on target
+			continue
+		}
+
+		updatedRules := target.updateIPs(lookupIPs)
+
+		// The IPs didn't change. No more to be done for this dnsName
+		if len(updatedRules) == 0 {
+			continue
+		}
+
+		// record which DNS names we used to update rules
+		updatedNames[dnsName] = lookupIPs
+
+		// accumulate rules we need to update with CIDR rules
+		for _, rule := range updatedRules {
+			affectedRules[rule] = append(affectedRules[rule], dnsName)
+		}
+	}
+
+	return affectedRules, updatedNames
+}
+
+// addRule places an api.Rule in the source list.
+// ruleKey must be a unique identifier for the sourceRule
+// newDNSNames and oldDNSNames indicate names that were newly added from this
+// rule, or that were seen in this rule but were already polled for.
+// If newDNSNames and oldDNSNames are both empty, the rule was not added to the
+// poll list.
+func (m *DNSPoller) addRule(ruleKey string, sourceRule *api.Rule) (newDNSNames, oldDNSNames []string) {
+	m.Lock()
+	defer m.Unlock()
+
+	for _, egressRule := range sourceRule.Egress {
+		for _, ToFQDN := range egressRule.ToFQDN {
+			for _, dnsName := range ToFQDN.FQDN {
+				target, dnsNameAlreadyExists := m.loadOrStore(dnsName)
+				target.addRule(ruleKey, sourceRule)
+
+				if dnsNameAlreadyExists {
+					oldDNSNames = append(oldDNSNames, dnsName)
+				} else {
+					newDNSNames = append(newDNSNames, dnsName)
+				}
+			}
+		}
+	}
+
+	return newDNSNames, oldDNSNames
+}
+
+// removeRule removes an api.Rule fomr the source rule set.
+// ruleKey must be a unique identifier for the sourceRule
+// isEmpty indicates no more rules rely on this DNS target
+func (m *DNSPoller) removeRule(ruleKey string, sourceRule *api.Rule) (noLongerPolled []string) {
+	m.Lock()
+	defer m.Unlock()
+
+	for _, egressRule := range sourceRule.Egress {
+		for _, ToFQDN := range egressRule.ToFQDN {
+			for _, dnsName := range ToFQDN.FQDN {
+
+				target, found := m.targets[dnsName]
+				if !found {
+					// no-op if nothing is there
+					return
+				}
+
+				isEmpty := target.removeRule(ruleKey, sourceRule)
+				if isEmpty {
+					noLongerPolled = append(noLongerPolled, dnsName)
+					delete(m.targets, dnsName)
+				}
+			}
+		}
+	}
+
+	return noLongerPolled
+}
+
+// loadOrStore returns the stored object for dnsName, or creates a new one if
+// none exists.
+func (m *DNSPoller) loadOrStore(dnsName string) (target *dnsPollTarget, loaded bool) {
+	target, loaded = m.targets[dnsName]
+	if !loaded {
+		target = &dnsPollTarget{
+			dnsName:     dnsName,
+			IPs:         make(map[string]bool),
+			sourceRules: make(map[string]*api.Rule),
+		}
+		m.targets[dnsName] = target
+	}
+
+	return target, loaded
+}
+
+// dnsPollTarget holds a DNS name that needs to be polled, its most recent
+// data, and the api.Rule that depends on it.
+type dnsPollTarget struct {
+	dnsName     string
+	IPs         map[string]bool
+	sourceRules map[string]*api.Rule // set of rules that use this DNS name
```

Comment:
I thought this was `map[FQDN]api.Rules`

I think you could have this split up. The dns pool target doesn't need to be aware of the existing rules, it only needs aware for which FQDN translates to which IPs. Then there's could be a different map that contains which `FQDN` belongs to which Rules.

If a `FQDN` no longer exists you can simply remove it from both maps.

edit: You might reply: "How can I known which rules are affected after doing a `updateIPs`? You can return `true` if `updateIPs` changed, the caller of `updateIPs` is aware of which FQDN you are currently doing the `updateIPs`. So if `updateIPs` returns true, the caller can simply return the value of `map[FQDN]api.Rules`

Reply:
I originally wrote all this with a `sync.Map` which meant everything had to live in one object. I can switch to what you're saying now, however. Note, though, that this is a different object than DNSPoller itself, so there are two layers here. What you're suggesting will mean the `dnsPollTarget` type isn't needed since I would just have two maps in DNSPoller: IPs `map[dnsName][]net.IP` and rules `map[dnsName][]*api.Rule`.